Create Durr_Hoyer_Post_1.md (#14)
qsharp-community/qsharp-community.github.io: _posts/Durr_Hoyer_Post_1.md

---

> title: "A quantum algorithm for finding the minimum of an unsorted table"

Reviewer: ```suggestion
title: "Building a Q# library to find table minima: Part 1"
```

> date:

Reviewer: We will bump this on the day it goes live, but it breaks the test builds if it is not there
```suggestion
date: 2020-06-22
```

categories:
  - Blog
tags:

>   - qsharp
>   - algorithm
>   - open source

Reviewer: These are tags to describe your post on the Q# community blog, I'd suggest some more descriptive ones below.

```suggestion
  - library
  - algorithm
  - in-development
```

---

> ### Durr Hoyer

Reviewer: ```suggestion
## Introduction
```

> Hello all! My name is Mridul Sarkar. I am an undergraduate at University of California Davis studying Scientific and Mathematical Computation. I began my journey from math to computers through my passion for ML algorithms. Eventually I stumbled upon quantum computing. I was instantly pulled in by the elegant algorithms and proposed efficiency.
> 
> 
> 
> ### Background
> Recently I began working on quantum computing with an idea to treat nodes of neural networks as qubits. From here I took a deep dive into the research papers and found myself amazed by the research being produced. I began with a Coursera course from St. Petersburg University on quantum computing with detail on quantum algorithm design and quantum computer architecture. From here I tinkered with Q# and developed basic algorithms that I learnt. I found Durr and Hoyer's paper and saw it as an ensemble of Duestchs, Shors, and Grovers Algorithm; a perfect stepping stone coming from my background. This project has pushed me and is still pushing me outside of my comfort zone. Learning Q# while slowly refining my quantum mechanics has been a rewarding experience. I would like to pass on what I have seen and understood in that time while open to feedback. I will likely be posting about this algorithm more as it continues developing and I learn more.

Reviewer: ```suggestion
Hello all! 
My name is Mridul Sarkar, and I am an undergraduate at University of California Davis studying Scientific and Mathematical Computation.
I wanted to share with the community my experiences in learning about the Dürr–Høyer algorithim and creating a library for Q# implementing it.
This project has pushed me and is still pushing me outside of my comfort zone, but learning Q# while slowly refining my quantum mechanics has been a rewarding experience. 
I would like to pass on what I have seen and understood in that time, as well as ask the community for feedback. 
I will likely be posting about this algorithm and library more as the library and my understanding continues developing.

### My Background
I began my journey from math to computers through my passion for ML algorithms, where I stumbled upon quantum computing. 
I was instantly pulled in by the elegant algorithms and proposed efficiency.
Recently, I started looking into quantum machine learning where you can model nodes of neural networks as qubits. 
I took a deep dive into the research papers and found myself amazed by what was out there.
I began with a Coursera course from St. Petersburg University on quantum computing with detail on quantum algorithm design and quantum computer architecture. 
From here I tinkered with Q# and developed basic algorithms that I previously learned.
I found Dürr and Høyer's paper and saw it as an fusion of Duestch's, Shor's, and Grover's Algorithm; a perfect next step for me.
```

### Motivation

> The motivation behind this project is to provide open source functionality. The efficiency that is proposed by this algorithm is Big-O = square root of the number of items in the table. Durr and Hoyer propose an ensemble of quantum algorithms in order to find the minimum of an unsorted table with a 50% success rate. This success rate is bounded by the Big-O. If we decide to alleviate this Big-O the probability of finding the minimum increases. With this comes some complexity in the inner workings of the library which will be explained when analyzing the algorithm. In order for this library to be used properly it must meet some guidelines. At the moment I am referencing the amazing template given by Dr. Sarah Kaiser https://github.com/crazy4pi314/qsharp-library-template in order to make this library usable in Q#. I am programming the algorithm in Q# with a python shell. The python shell will be provided below. The circuit's gates will be analyzed and some code is given.

Reviewer: - Can you maybe try to explain what Big-O is or without using it? Not all the readers may be familiar/will need a refresher :)
- Link to your repo early and often
- Can you clarify what you mean by a Python shell? Do you mean a Python host for Q#?

> ### The Algorithm

Reviewer: ```suggestion
## The Dürr–Høyer Algorithm
```

> A brief introduction to the Durr and Hoyer Algorithm can be derived from the Quantum Minimum Searching Algorithm outlined in 'A quantum algorithm for finding the minimum' [(1)]:

Reviewer: Can you give a brief summary for those that won't go read the whole paper?

Durr and Hoyer propose a computationally efficient routine to find the minimum of a table of distinct unsorted integers. The algorithm is as follows:

![QMSA](/assets/images//DurrHoyer-QMSA.JPG "QMSA")

> #### Psuedocode

Reviewer: Can you maybe describe the steps in words instead? It's not super accessible for people using screen readers to have some of the most important info in the post. You don't have to worry about the Python then.

```
    t = table.flatten()
    N = len(t)
    y = randomuniform[1..N] #N is a integer
    time_limit = (22.5*sqrt(N)=1.4*log(N)^2)
    q = preprocess.simulate(t,y)

    while time.clock() < time_limit:
        for i in range(N):
            while t[i] < t[y]:
                i = N-i
                if N is even
                    q=Algorithm with Hadamard gates[0...j]
                    y'=q[0]
                else
                    t=Algorithm with QFT instead of H[0...j]
                    y'=q[0]
        if t[y'] < t[y]:
            y = y'
            return
    return y
```

> ------------------

Reviewer: ```suggestion

```

> It is easy to understand the conceptual implementation in python. It is important to note steps:
> ### (a,b)

Reviewer: ```suggestion
It is easy to understand the conceptual implementation in python. It is important to note steps a and b:
```

> ------------------

Reviewer: ```suggestion

```

> ![a](/assets/images//a.JPG)   

Reviewer: This only seems to be step a?

For now, the more prevalent question is what sort of initialization we can use to verify the integrity and efficiency of this algorithm. My most immediate thought is to entangle all the qubits and superimpose them. The classical value of the integer is not represented here, only the index of the integer is captured at a quantum level on the table assuming an input datatype of BigEndian. If the goal is to test the algorithm's probabilistic integrity this plan of action will suffice.

The register can be initiated as follows:

![latex_imp](/assets/images//latex_information.JPG)

Obviously this is unachievable on current quantum computer architecture. We only have a qubit, the above example implies we have harnessed the tritbit and so on till n bit. In reality we will register all integers as unique binary strings which will equate to their qubits.
 
Luckily PrepareUniformSuperposition() in Q# does this for us easily. I have implemented it as follows:

> ![intialize](/assets/images//intialize.JPG)

Reviewer: I would strongly suggest looking at the [Q# style guide](https://docs.microsoft.com/en-us/quantum/contributing/style-guide?tabs=guidance) and making this an actual fenced code block in the text not a screen grab.

> Note: make sure to use LittleEndian() on the register of qubits before applying PrepareUniformSuperposition(). From here you will have to use LittleEndianAsBigEndian() to set the right data type after using PrepareUniformSuperposition() as it spits out LE datatype. I didn't include this so it looked neat.

Reviewer: You have to include the `LittleEndian` or this code is not valid.
